### ISSUE TITLE
Nested VJP function always triggers recompilation 

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#3755]
I'm sorry to report that my last contribution https://github.com/google/jax/issues/3667 was not a perfect fix for the issue.  While it does seem to ensure that the returned cotangent function is a pytree, it does not ensure that regenerating the cotangent function produces an equivalent pytree.  Consider the new test at the bottom of this snippet:
```python
  def test_nestable_vjp(self):
    # Verify that https://github.com/google/jax/issues/3667 is resolved.
    def f(x):
        return x ** 2

    @api.custom_vjp
    def g(x):
        return f(x)

    def g_fwd(x):
        y, f_vjp = api.vjp(f, x)
        return y, f_vjp

    def g_bwd(f_vjp, y_bar):
        return f_vjp(y_bar)

    g.defvjp(g_fwd, g_bwd)

    # Check that VJP can be nested in simple situations.  For this to pass,
    # vjp has to return a PyTree.
    _, g_vjp = api.vjp(g, 1.0)
    y, = g_vjp(1.0)
    self.assertAllClose(y, jnp.array(2.0))

    # Check that VJP can be nested in complex situations.  For this to pass,
    # vjp can't treat the closed-over tracer x as a static argument.
    @jit
    def z(x):
        _, g_vjp = api.vjp(g, x)
        return g_vjp
    y, = z(1.0)(3.0)
    self.assertAllClose(y, jnp.array(6.0))

    # Check that VJP functions don't induce recompilation.
    compiled = 0
    @jit
    def w(this_g_vjp):
        nonlocal compiled
        compiled += 1
    _, g_vjp = api.vjp(g, 1.0)
    w(g_vjp)
    _, g_vjp = api.vjp(g, 1.0)
    w(g_vjp)
    _, g_vjp = api.vjp(g, 2.0)
    w(g_vjp)
    assert compiled == 1
```
Currently, `compiled` is 3.  This is because there are `partial` objects buried in the pytree, and upon regeneration of `g_vjp`, we get different objects, which hash differently.

In trying to fix this, I tried to replace the pattern `Partial(partial(` in the code with an improved `Partial` class from [here](https://github.com/google/jax/issues/3667#issuecomment-656240407).  The problem improves, but then `Jaxpr` is not a pytree.

Long story short, there are two possible ways of moving forward:

- Try to work around the problem by not regenerating cotangent functions.  Generate them exactly once.
- Regfister a number of JAX's internal classes as pytrees, including `Jaxpr`, `Var`, and `Atom`.

I'll probably try just working around this for now, but I was wondering what your thoughts are on ultimately doing the registration?

attn: @shoyer @mattjj

**Comment by shoyer** [comment]
> * Regfister a number of JAX's internal classes as pytrees, including `Jaxpr`, `Var`, and `Atom`.

That sound rather invasive!

Is there a reason why you can't put the whole thing -- including the call to `api.vjp` and the use of returned jvp function -- under a single `jit` decorated function?

**Comment by shoyer** [comment]
> Right, that's what I had in mind to ensure that the cotangent function is not regenerated. It just feels weird to have a jitted function (say `f`) that returns different pytrees (as far as jit is concerned) when it is called with the exact same arguments.

Well, you're calling `api.vjp()` on `f`, which is explicitly invoking JAX's auto-diff machinery. In most cases, that machinery has cost that is comparable to the cost of `jit` compilation, so even if this were possible I'm not sure if it would make a meaningful difference.

**Comment by shoyer** [comment]
I'm going to close this as "working as intended" for now, but feel free to open a new issue if this comes up in a new context.
